### PR TITLE
up(linter): add rule fixer

### PR DESCRIPTION
**Pull request by DonIsaac**
Adds a new `RuleFixer` struct that gets passed to `ctx.diagnostic_with_fix`.

This PR is half proposal and half implementation. I'd love thoughts and feedback on its design (including whether or not it adds enough benefit to make it worth merging this PR).

### CODE REVIEW

File: crates/oxc_span/src/span.rs
Comment (Boshen) [inline crates/oxc_span/src/span.rs:1]: Regarding all these changes, we should have a separate PR copying all the methods from https://docs.rs/text-size/1.1.1/text_size/struct.TextRange.html

File: crates/oxc_ast/src/span.rs [hunk 1]
```diff
@@ -2,6 +2,33 @@ use oxc_span::{GetSpan, Span};
 
 use crate::ast::*;
 
+macro_rules! impl_get_span {
```
Comment (Boshen) [inline crates/oxc_ast/src/span.rs:5]: Let's have as less macro as possible in these crates, a simple copy paste is totally fine here.

I have a rule https://oxc.rs/docs/contribute/rules.html#development-policy

> Avoid heavy macros or generics or any Rust techniques that would penalize compilation speed.

It gets heavy when we repeat these.

**Review by Boshen — COMMENTED** [review]
The direction is good, we should just port most of eslint 😅

File: crates/oxc_span/src/span.rs
Comment (DonIsaac) [reply crates/oxc_span/src/span.rs:1]: https://github.com/oxc-project/oxc/pull/3592

File: crates/oxc_ast/src/span.rs (hunk 1 above)
Comment (DonIsaac) [reply crates/oxc_ast/src/span.rs]: Removed

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Definitely a better API!

**Pull request merged**